src/ship/GunManager.cpp

```diff
@@ -0,0 +1,519 @@
+			weapon.group--;
+		}
+	}
+
+	// Finally, clear the group from the list
+	m_groups.erase(m_groups.begin() + group);
+}
+
+const Body *GunManager::GetGroupTarget(uint32_t group)
+{
+	if (group >= m_groups.size())
+		return nullptr;
+
+	return m_groups[group].target;
+}
+
+void GunManager::SetGroupTarget(uint32_t group, const Body *target)
+{
+	if (group >= m_groups.size())
+		return;
+
+	GroupState &gs = m_groups[group];
+	// No weapons at all, ignore
+	if (!gs.weapons.any())
+		return;
+
+	// Clear the prior callback, if any
+	if (gs.target)
+		gs.onTargetDestroyed.disconnect();
+
+	m_groups[group].target = target;
+
+	if (target) {
+		gs.onTargetDestroyed = target->onDelete.connect([=]() {
+			this->SetGroupTarget(group, nullptr);
+		});
+	}
+}
+
+void GunManager::SetGroupFiring(uint32_t group, bool firing)
+{
+	if (group >= m_groups.size())
+		return;
+
+	GroupState &gs = m_groups[group];
+	// No weapons at all, ignore
+	if (!gs.weapons.any())
+		return;
+
+	gs.firing = firing;
+	m_isAnyFiring |= firing;
+
+	if (firing) {
+		// Update the next-firing time for all weapons in this group so they actually fire
+		// (Ensures you can't spam the fire command to shoot more rapidly than the weapon's rate of fire)
+		for (size_t idx = 0; idx < m_weapons.size(); idx++) {
+			if (gs.weapons[idx]) {
+				m_weapons[idx].nextFireTime = std::max(m_weapons[idx].nextFireTime, Pi::game->GetTime());
+			}
+		}
+	} else {
+		m_stoppedNextFrame |= gs.weapons;
+	}
+}
+
+void GunManager::SetGroupFireWithoutTargeting(uint32_t group, bool enabled)
+{
+	if (group >= m_groups.size())
+		return;
+
+	GroupState &gs = m_groups[group];
+	gs.fireWithoutTargeting = enabled;
+}
+
+vector3d GunManager::GetGroupLeadPos(uint32_t group)
+{
+	if (group >= m_groups.size() || !m_groups[group].target || m_groups[group].weapons.none())
+		return vector3d(0, 0, 0);
+
+	GroupState &gs = m_groups[group];
+	double inv_count = 1.0 / double(gs.weapons.count());
+	vector3d lead_pos = vector3d(0, 0, 0);
+
+	for (size_t idx = 0; idx < m_weapons.size(); idx++) {
+		if (gs.weapons[idx]) {
+			lead_pos += m_weapons[idx].currentLeadPos * inv_count;
+		}
+	}
+
+	return lead_pos;
+}
+
+float GunManager::GetGroupTemperatureState(uint32_t group)
+{
+	if (group >= m_groups.size() || m_groups[group].weapons.none())
+		return 0.f;
+
+	GroupState &gs = m_groups[group];
+	double inv_count = 1.0 / double(gs.weapons.count());
+	float temperature = 0.f;
+
+	for (size_t idx = 0; idx < m_weapons.size(); idx++) {
+		if (gs.weapons[idx]) {
+			temperature += inv_count * (m_weapons[idx].temperature / m_weapons[idx].data.overheatThreshold);
+		}
+	}
+
+	return temperature;
+}
+
+void GunManager::StaticUpdate(float deltaTime)
+{
+	bool isAnyFiring = false;
+
+	static ImGuiOnceUponAFrame thisFrame;
+	bool shouldDrawDebug = m_parent && m_parent->IsType(ObjectType::PLAYER) && thisFrame;
+	if (shouldDrawDebug)
+		ImGui::Begin("GunMangerDebug", nullptr, ImGuiWindowFlags_NoFocusOnAppearing);
+
+	if (shouldDrawDebug) {
```

Comment:
Duplication of check..
1. Combine checks
2. Style suggestion (safety) _always_ use braces for compound statements

Reply:
Actually removed in a later commit.

Note to self: don't do per-commit reviews :P